client/my-sites/term-tree-selector/add-term.jsx

```diff
@@ -0,0 +1,232 @@
+	}
+
+	openDialog( event ) {
+		event.preventDefault();
+
+		this.setState( {
+			showDialog: true,
+			selectedParent: []
+		} );
+	}
+
+	closeDialog() {
+		this.setState( this.constructor.initialState() );
+	}
+
+	onParentChange( item ) {
+		this.setState( {
+			selectedParent: [ item.ID ],
+			isTopLevel: false
+		}, this.isValid );
+	}
+
+	onTopLevelChange() {
+		this.setState( {
+			isTopLevel: ! this.state.isTopLevel,
+			selectedParent: []
+		}, this.isValid );
+	}
+
+	getSelectedValues() {
+		const termName = ReactDom.findDOMNode( this.refs.termName ).value.trim();
+		const parent = this.state.selectedParent.length ? this.state.selectedParent[ 0 ] : 0;
+
+		return {
+			name: termName,
+			parent: parent
+		};
+	}
+
+	isValid() {
+		let error;
+
+		const values = this.getSelectedValues();
+
+		if ( ! values.name.length ) {
+			error = true;
+		}
+
+		if ( find( this.props.terms, values ) ) {
+			error = this.props.translate( 'Name already exists', {
+				context: 'Terms: Add term error message - duplicate term name exists',
+				textOnly: true
+			} );
+		}
+
+		if ( error !== this.state.error ) {
+			this.setState( {
+				error: error,
+				isValid: ! error
+			} );
+		}
+
+		return ! error;
+	}
+
+	validateInput( event ) {
+		if ( 13 === event.keyCode ) {
+			this.saveTerm();
+		} else {
+			this.isValid();
+		}
+	}
+
+	saveTerm() {
+		const values = this.getSelectedValues();
+		if ( ! this.isValid() ) {
+			return;
+		}
+
+		const { siteId, taxonomy } = this.props;
+
+		this.props.addTerm( siteId, taxonomy, values );
+		this.closeDialog();
+	}
+
+	render() {
+		const { labels, siteId, taxonomy, translate } = this.props;
+		const buttons = [ {
+			action: 'cancel',
+			label: translate( 'Cancel' )
+		}, {
+			action: 'add',
+			label: translate( 'Add' ),
+			isPrimary: true,
+			disabled: ! this.state.isValid,
+			onClick: this.boundSaveTerm
+		} ];
+
+		const { searchTerm, selectedParent } = this.state;
+		const query = {};
+		if ( searchTerm && searchTerm.length ) {
+			query.search = searchTerm;
+		}
+
+		const isError = this.state.error && this.state.error.length;
+
+		return (
+			<div className="term-tree-selector__add-term">
+				<Button borderless compact={ true } onClick={ this.boundOpenDialog }>
+					<Gridicon icon="folder" /> { labels.add_new_item }
+				</Button>
+				<Dialog
+					autoFocus={ false }
+					isVisible={ this.state.showDialog }
+					buttons={ buttons }
+					onClose={ this.boundCloseDialog }
+					additionalClassNames="term-tree-selector__add-term-dialog">
+					<FormSectionHeading>{ labels.add_new_item }</FormSectionHeading>
+					<FormFieldset>
+						<FormTextInput
+							autoFocus={ this.state.showDialog && ! viewport.isMobile() }
+							placeholder={ labels.new_item_name }
+							ref="termName"
+							isError={ isError }
+							onKeyUp={ this.boundValidateInput } />
+						{ isError ? <FormInputValidation isError={ true } text={ this.state.error } /> : null }
```

Comment:
As an FYI you can always omit `={ true }` e.g. `<FormInputValidation isError />`